### PR TITLE
feat: adds support for detecting nested trait usage for SoftDeletes trait

### PR DESCRIPTION
**Pull request by IAmRGroot**
- [x] Added or updated tests
- [x] Documented user facing changes
- [ ] Updated CHANGELOG.md

Related to #837 

**Changes**

- Use recursive `$modelReflection->getTraits(true)` instead of `$modelReflection->hasTraitUse()`

Tests:
- Added `NestedSoftDeletes` trait
- Groups model now uses `NestedSoftDeletes` instead of `SoftDeletes`
- Added `->withThrashed` to the Users Group relation to add a test case

**Breaking changes**

None

### CODE REVIEW

**Comment by szepeviktor** [comment]
Spotted a `rgroot`email in your commit.

**Comment by szepeviktor** [comment]
I think Larastan's SoftDeletes detection needs to be fixed.

**Comment by IAmRGroot** [comment]
@szepeviktor thanks, it should be fixed now!

**Comment by IAmRGroot** [comment]
@szepeviktor please let me know if you need anything from my end

**Review by szepeviktor — APPROVED** [review]

**Comment by IAmRGroot** [comment]
@szepeviktor I've changed it to use the new phpstan recursive getTraits function. Please let me know if I need to change anything

File: composer.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     ],
     "require": {
         "php": "^7.2 || ^8.0",
+        "ext-json": "^8.0",
```
Comment (canvural) [inline composer.json:15]: Can you revert this change? As you can see from the GA, it's causing composer to fail.

Comment (szepeviktor) [reply composer.json:15]: It could be `"ext-json": "*",`

**Review by szepeviktor — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (canvural) [reply composer.json:15]: It was like that already.

Comment (szepeviktor) [reply composer.json:15]: https://github.com/nunomaduro/larastan/blob/master/composer.json#L26

😲

Comment (IAmRGroot) [reply composer.json:15]: Ah yes I thought it only changed the order. I'll revert it

**Comment by canvural** [comment]
Thank you!

File: tests/Application/app/User.php [hunk 2]
```diff
@@ -63,7 +63,7 @@ public function scopeActive(Builder $query): Builder
     /** @phpstan-return BelongsTo<Group, User> */
     public function group(): BelongsTo
     {
-        return $this->belongsTo(Group::class);
+        return $this->belongsTo(Group::class)->withTrashed();
```
Comment (nunomaduro) [inline tests/Application/app/User.php:66]: I feel that you should add a new test, and not modify an existing one. cc @canvural

Comment (nunomaduro) [reply tests/Application/app/User.php:66]: Ups, looks like @canvural just merged this. Can you confirm that modifying an existing test is normal?

Comment (canvural) [reply tests/Application/app/User.php:66]: Actually this not a test case. It's just the dummy application that the tests are using. 

So now I think of it, this actually misses a test case 😅 Thank you @nunomaduro 

@IAmRGroot Can you add a new method to https://github.com/nunomaduro/larastan/blob/master/tests/Features/Models/Relations.php called `testNestedSoftDeletesTrait` maybe, that uses the `group` method on `User` class. And return type should be `BelongsTo<Group, User>` You can check how other methods are written in that file.

Comment (szepeviktor) [reply tests/Application/app/User.php:66]: What a turmoil!

Comment (IAmRGroot) [reply tests/Application/app/User.php:66]: I did initially change the test application to trigger a check for the nested SoftDeletes trait.

Maybe in hindsight this was not needed as the existing tests `testWithTrashedWithBelongsToRelation`, `testOnlyTrashedWithBelongsToRelation`, `testWithoutTrashedWithBelongsToRelation` already check the `$user->group()->withTrashed()` function:
https://github.com/nunomaduro/larastan/blob/master/tests/Features/Models/Relations.php#L185
